subprojects/publish/src/main/java/org/gradle/api/publish/internal/mapping/VariantDependencyResolver.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.publish.internal.mapping;
+
+import org.gradle.api.artifacts.DependencyConstraint;
+import org.gradle.api.artifacts.ModuleDependency;
+import org.gradle.api.artifacts.ModuleVersionIdentifier;
+import org.gradle.api.publish.internal.validation.VariantWarningCollector;
+
+import javax.annotation.Nullable;
+
+/**
+ * Given declared dependencies and dependency constraints, determines the coordinates
+ * that should be used to reference the dependency in published metadata. Instances
+ * of this interface are scoped to a given published variant, as each variant may
+ * have different rules for resolving published dependencies.
+ *
+ * <p>This resolver can resolve dependencies to either component or variant precision.
+ * When resolving to component precision, the returned coordinates will reference the
+ * root coordinates of a multi-coordinate publication. When resolving to variant precision,
+ * the returned coordinates will reference the child coordinates of the variant selected by
+ * the resolved dependency.</p>
+ */
+public interface VariantDependencyResolver {
+
+    /**
+     * Determines the published coordinates for a dependency to variant-level precision.
+     *
+     * @throws RuntimeException If {@code dependency} is a project dependency and the project cannot be resolved.
```

Comment:
❓ Where does this throw the `RuntimeException`?  Doesn't it just use the warning collector?  I don't think this should be documenting runtime exceptions that potentially come from several levels down in the call stack.

Reply:
💭 It also seems like the `VariantDependencyResolver` should be created with the warning collector as a gettable field, and not have to pass it to each of these resolve call.

Reply:
I think we should definitely be documenting this exception. If the ID of the provided project dependency does not point to a project, this method will throw an exception. I think that is definitely behavior that should be documented.  

This was existing behavior of the code. I'd prefer to avoid changing behavior in this PR, though we can discuss if this is something we want to allow. 

In general, I think it makes more sense to throw an exception if a user depends on a project that does not exist, rather than change it to a warning.

Reply:
Agreed. I've updated the javadoc of `ProjectDependencyPublicationResolver` to document the exception thrown